### PR TITLE
Support Suppression

### PR DESCRIPTION
**Pull request by pseudomuto**
@quelledanielle and @raulriera, mind taking a :eyes:?
# The Problem

It's just bad form to pester a user with update notifications. Currently, the approach is to check for update status during `applicationDidBecomeActive`. 

This poses a problem. Imaging the following scenario:
- User opens the app and sees a recommended update notice
- Then she gets a text messsage
- She comes back to the app (triggering `applicationDidBecomeActive` again) and sees another update notice

:hankey: 
# The Solution

Add the ability to suppress update checks for a configurable period of time. This is accomplished by calling `Retired.suppressUntil` and passing in a date object.

When `Retired.check` is called, the call will be skipped if the suppression date is greater than the current date.
- [x] Tests passing
- [x] README updated
- [x] Example app updated

### CODE REVIEW

File: Examples/Retired iOS Demo/Retired iOS Demo/AppDelegate.swift [hunk 1]
```diff
@@ -19,11 +21,12 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
   var window: UIWindow?
 
   func application(application: UIApplication, didFinishLaunchingWithOptions launchOptions: [NSObject: AnyObject]?) -> Bool {
+    Retired.configure(versionURL)
     return true
   }
 
   func applicationDidBecomeActive(application: UIApplication) {
-    Retired.check(versionURL) { updateRequired, message, error in
+    try! Retired.check() { updateRequired, message, error in
```
Comment (raulriera) [inline Examples/Retired iOS Demo/Retired iOS Demo/AppDelegate.swift:29]: Danger! :scream:

File: Examples/Retired iOS Demo/Retired iOS Demo/AppDelegate.swift [hunk 2]
```diff
@@ -50,4 +53,8 @@ extension Message {
   private func goToAppStore(action: UIAlertAction) {
     UIApplication.sharedApplication().openURL(iTunesURL)
   }
+
+  private func doItAgainLater(action: UIAlertAction) {
+    Retired.suppressUntil(NSDate(timeIntervalSinceNow: intervalBetweenRequests))
+  }
```
Comment (raulriera) [inline Examples/Retired iOS Demo/Retired iOS Demo/AppDelegate.swift]: Shouldn't this be built into the library?

File: Sources/DownloadService.swift [hunk 3]
```diff
@@ -22,7 +22,7 @@ private let session: NSURLSession = {
   return NSURLSession(configuration: config)
 }()
 
-class DownloadService {
+internal class DownloadService {
```
Comment (raulriera) [inline Sources/DownloadService.swift]: FYI, internal is the default permission. So no need to mark them when using .frameworks

File: Sources/Fetcher.swift [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+public protocol FileFetcher {
+  func check(completion: RetiredCompletion)
+}
+
+class Fetcher: FileFetcher {
+  let url: NSURL
+  let bundle: NSBundle
+
+  init(_ url: NSURL, bundle: NSBundle) {
```
Comment (raulriera) [inline Sources/Fetcher.swift]: why `_`? Kind of against init conventions

File: Sources/StoredSetting.swift [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        settings.setObject(setting, forKey: key)
+      } else {
+        settings.removeObjectForKey(key)
+      }
+
+      settings.synchronize()
+    }
+  }
+
+  public init(_ name: String, container: NSUserDefaults = NSUserDefaults.standardUserDefaults()) {
```
Comment (raulriera) [inline Sources/StoredSetting.swift]: Same here 😔 named arguments are a great thing about Swift and Objective-C

File: Examples/Retired iOS Demo/Retired iOS Demo/AppDelegate.swift (hunk 1 above)
Comment (pseudomuto) [inline Examples/Retired iOS Demo/Retired iOS Demo/AppDelegate.swift:29]: Demo :smile:

File: Tests/StoredSettingTests.swift [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  }()
+
+  override func tearDown() {
+    super.tearDown()
+
+    settingsContainer.removeObjectForKey(settingKey)
+    settingsContainer.synchronize()
+  }
+
+  func test_Get_Value_Pulls_From_Container() {
```
Comment (raulriera) [inline Tests/StoredSettingTests.swift]: Ruby 😂

File: Examples/Retired iOS Demo/Retired iOS Demo/AppDelegate.swift (hunk 2 above)
Comment (pseudomuto) [inline Examples/Retired iOS Demo/Retired iOS Demo/AppDelegate.swift]: I thought about that. To be clear, you're suggesting adding an interval to the configure call (with a default) and just setting it every time a successful fetch is made?

File: Tests/StoredSettingTests.swift (hunk 6 above)
Comment (pseudomuto) [inline Tests/StoredSettingTests.swift]: The underscores?

File: Examples/Retired iOS Demo/Retired iOS Demo/AppDelegate.swift (hunk 2 above)
Comment (raulriera) [inline Examples/Retired iOS Demo/Retired iOS Demo/AppDelegate.swift]: Yeah, making the "retry" thing configurable. So it handles it itself whenever a user presses cancel

File: Tests/StoredSettingTests.swift (hunk 6 above)
Comment (raulriera) [inline Tests/StoredSettingTests.swift]: Yeah, snake_case + camelCase together looks pretty weird

File: Examples/Retired iOS Demo/Retired iOS Demo/AppDelegate.swift (hunk 2 above)
Comment (dphm) [inline Examples/Retired iOS Demo/Retired iOS Demo/AppDelegate.swift]: Sounds good, I was also expecting the interval to be configurable.

File: Tests/StoredSettingTests.swift (hunk 6 above)
Comment (pseudomuto) [inline Tests/StoredSettingTests.swift]: It really does...I'll update

**Comment by raulriera** [comment]
FYI in [Version.swift](https://github.com/pseudomuto/Retired/blob/master/Sources/Version.swift#L24) you could use [RawRepresentable](https://developer.apple.com/library/mac/documentation/Swift/Reference/Swift_RawRepresentable_Protocol/index.html) instead to init enums

File: Tests/StoredSettingTests.swift (hunk 6 above)
Comment (dphm) [inline Tests/StoredSettingTests.swift]: 😛  +1 :snake: :camel:

**Comment by raulriera** [comment]
BTW, love the idea of "personal libraries" to handle things we know we are going to need for future projects

File: Tests/RetiredTests.swift [hunk 7]
```diff
@@ -6,84 +6,45 @@
...
-      XCTAssertEqual(Retired.errorCodeNotFound, error!.code)
-    }
+  func test_Check_When_Not_Supressed_Calls_The_Fetcher() {
+    try! Retired.check() { _, _, _ in }
+    XCTAssertTrue(fetcher.called)
   }
 
-  func test_Check_When_Error_Occurs() {
-    stubRequest("GET", versionURL).andFailWithError(NSError(domain: "err", code: 1, userInfo: nil))
+  func test_Check_When_Supressed_Skips_The_Call() {
```
Comment (dphm) [inline Tests/RetiredTests.swift]: Did you also test check after suppression interval calls the fetcher?

**Comment by dphm** [comment]
> Did you also test check after suppression interval calls the fetcher?

Making sure my comment doesn't get lost

**Comment by raulriera** [comment]
Good

**Comment by pseudomuto** [comment]
Thanks for the review guys!

**Pull request merged**